File: launcher/src/test/scala/launcher/integration/IntegrationTests.scala
package launcher.integration

import application.dto.{BalanceDTO, CreateAccountDTO, CreateMovementDTO, TransferMovementDTO}
import domain.model.MovementDescription.{CREDIT_TRANSFER_DESCRIPTION, DEBIT_TRANSFER_DESCRIPTION}
import domain.model.{Account, Movement, MovementType}
import launcher.FinancialApplication
import org.junit.jupiter.api.*
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.reactive.AutoConfigureWebTestClient
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.MediaType
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.web.reactive.server.WebTestClient
import org.springframework.web.reactive.function.BodyInserters

import java.math.{BigDecimal, RoundingMode}
import java.util
import java.util.UUID
import java.util.concurrent.atomic.AtomicReference
import java.util.stream.IntStream

object IntegrationTests {
  val accountId1 = new AtomicReference[String]
  val accountId2 = new AtomicReference[String]
  val expectedAmount: BigDecimal = BigDecimal.TEN.setScale(2, RoundingMode.DOWN)
  val expectedZero: BigDecimal = BigDecimal.ZERO.setScale(2, RoundingMode.DOWN)
}

@SpringBootTest(
  properties = Array(
    "spring.r2dbc.url=r2dbc:h2:mem:///./testdb",
    "spring.liquibase.url=jdbc:h2:mem:./testdb;DB_CLOSE_DELAY=-1"
  )
)
@AutoConfigureWebTestClient
@ContextConfiguration(classes = Array(classOf[FinancialApplication]))
@TestMethodOrder(
  classOf[MethodOrderer.OrderAnnotation]
)
class IntegrationTests {

  @Autowired private val webClient: WebTestClient = null

  @Order(1)
  @Test def createAccountsWithSuccess(): Unit = {
    val accountId1id = webClient.post
      .uri("/api/v1/accounts")
      .contentType(MediaType.APPLICATION_JSON)
      .body(BodyInserters.fromValue(createAccountDTO))
      .exchange
      .expectStatus
      .isCreated
      .expectBody(classOf[Account])
      .returnResult
      .getResponseBody
      .id
    IntegrationTests.accountId1.set(accountId1id)

    val accountId2Id = webClient.post
      .uri("/api/v1/accounts")
      .contentType(MediaType.APPLICATION_JSON)
      .body(BodyInserters.fromValue(createAccountDTO))
      .exchange
      .expectStatus
      .isCreated
      .expectBody(classOf[Account])
      .returnResult
      .getResponseBody
      .id
    IntegrationTests.accountId2.set(accountId2Id)
  }

  @Order(2)
  @Test def createDepositsWithSuccess(): Unit = {
    webClient.post
      .uri("/api/v1/movements/" + IntegrationTests.accountId1.get + "/deposit")
      .contentType(MediaType.APPLICATION_JSON)
      .body(BodyInserters.fromValue(createMovementDTO))
      .exchange
      .expectStatus
      .is2xxSuccessful
      .expectBody(classOf[Movement])

    webClient.post
      .uri("/api/v1/movements/" + IntegrationTests.accountId2.get + "/deposit")
      .contentType(MediaType.APPLICATION_JSON)
      .body(BodyInserters.fromValue(createMovementDTO))
      .exchange
      .expectStatus
      .is2xxSuccessful
      .expectBody(classOf[Movement])
      .returnResult()
  }

  @Order(3)
  @Test def createWithdrawsWithSuccess(): Unit = {
    webClient.post
      .uri("/api/v1/movements/" + IntegrationTests.accountId2.get + "/withdraw")
      .contentType(MediaType.APPLICATION_JSON)
      .body(BodyInserters.fromValue(createMovementDTO))
      .exchange
      .expectStatus
      .is2xxSuccessful
      .expectBody(classOf[Movement])

    webClient.post
      .uri("/api/v1/movements/" + IntegrationTests.accountId1.get + "/withdraw")
      .contentType(MediaType.APPLICATION_JSON)
      .body(BodyInserters.fromValue(createMovementDTO))
      .exchange
      .expectStatus
      .is2xxSuccessful
      .expectBody(classOf[Movement])
      .returnResult()
  }

  @Order(4)
  @Test def retrieveCreditMovements(): Unit = {
    val account1Movements = webClient.get
      .uri("/api/v1/movements/" + IntegrationTests.accountId1.get + "/credit")
      .exchange
      .expectStatus
      .is2xxSuccessful
      .expectBody(classOf[Array[Movement]])
      .returnResult
      .getResponseBody

    val account2Movements = webClient.get
      .uri("/api/v1/movements/" + IntegrationTests.accountId2.get + "/credit")
      .exchange
      .expectStatus
      .is2xxSuccessful
      .expectBody(classOf[Array[Movement]])
      .returnResult
      .getResponseBody

    Assertions.assertEquals(1, account1Movements.length)
    Assertions.assertEquals(1, account2Movements.length)
    Assertions.assertEquals(
      IntegrationTests.expectedAmount,
      account1Movements(0).amount
    )
    Assertions.assertEquals(
      IntegrationTests.expectedAmount,
      account2Movements(0).amount
    )
  }

  @Order(5)
  @Test def retrieveDebitMovements(): Unit = {
    val account1Movements = webClient.get
      .uri("/api/v1/movements/" + IntegrationTests.accountId1.get + "/debit")
      .exchange
      .expectStatus
      .is2xxSuccessful
      .expectBody(classOf[Array[Movement]])
      .returnResult
      .getResponseBody

    val account2Movements = webClient.get
      .uri("/api/v1/movements/" + IntegrationTests.accountId2.get + "/debit")
      .exchange
      .expectStatus
      .is2xxSuccessful
      .expectBody(classOf[Array[Movement]])
      .returnResult
      .getResponseBody

    Assertions.assertEquals(1, account1Movements.length)
    Assertions.assertEquals(1, account2Movements.length)
    Assertions.assertEquals(
      IntegrationTests.expectedAmount,
      account1Movements(0).amount
    )
    Assertions.assertEquals(
      IntegrationTests.expectedAmount,
      account2Movements(0).amount
    )
  }

  @Order(6)
  @Test def assertAccountsBalances(): Unit = {
    val account1Balance = webClient.get
      .uri("/api/v1/accounts/" + IntegrationTests.accountId1.get + "/balance")
      .exchange
      .expectStatus
      .is2xxSuccessful
      .expectBody(classOf[BalanceDTO])
      .returnResult
      .getResponseBody

    val account2Balance = webClient.get
      .uri("/api/v1/accounts/" + IntegrationTests.accountId2.get + "/balance")
      .exchange
      .expectStatus
      .is2xxSuccessful
      .expectBody(classOf[BalanceDTO])
      .returnResult
      .getResponseBody

    Assertions.assertEquals(
      IntegrationTests.expectedZero,
      account1Balance.balance
    )
    Assertions.assertEquals(
      IntegrationTests.expectedZero,
      account2Balance.balance
    )
  }

  @Order(7)
  @Test def assertTransfersWorkWithSuccess(): Unit = {
    // deposit 10 bucks in each account
    createDepositsWithSuccess()
    val transferMovement = new TransferMovementDTO(
      IntegrationTests.accountId2.get,
      IntegrationTests.expectedAmount
    )

    webClient.post
      .uri("/api/v1/movements/" + IntegrationTests.accountId1.get + "/transfer")
      .contentType(MediaType.APPLICATION_JSON)
      .body(BodyInserters.fromValue(transferMovement))
      .exchange
      .expectStatus
      .is2xxSuccessful
      .expectBody(classOf[Movement])

    val account1Balance = webClient.get
      .uri("/api/v1/accounts/" + IntegrationTests.accountId1.get + "/balance")
      .exchange
      .expectStatus
      .is2xxSuccessful
      .expectBody(classOf[BalanceDTO])
      .returnResult
      .getResponseBody

    val account2Balance = webClient.get
      .uri("/api/v1/accounts/" + IntegrationTests.accountId2.get + "/balance")
      .exchange
      .expectStatus
      .is2xxSuccessful
      .expectBody(classOf[BalanceDTO])
      .returnResult
      .getResponseBody

    Assertions.assertEquals(
      IntegrationTests.expectedAmount.add(IntegrationTests.expectedAmount),
      account2Balance.balance
    )
    Assertions.assertEquals(
      IntegrationTests.expectedZero,
      account1Balance.balance
    )

    val account2CreditMovements = webClient.get
      .uri("/api/v1/movements/" + IntegrationTests.accountId2.get + "/credit")
      .exchange
      .expectStatus
      .is2xxSuccessful
      .expectBody(classOf[Array[Movement]])
      .returnResult
      .getResponseBody

    val account1DebitMovements = webClient.get
      .uri("/api/v1/movements/" + IntegrationTests.accountId1.get + "/debit")
      .exchange
      .expectStatus
      .is2xxSuccessful
      .expectBody(classOf[Array[Movement]])
      .returnResult
      .getResponseBody

    val firstAccount2CreditMovement = account2CreditMovements(0)
    val firstAccount1DebitMovement = account1DebitMovements(0)

    Assertions.assertEquals(
      firstAccount1DebitMovement.amount,
      firstAccount2CreditMovement.amount
    )
    Assertions.assertEquals(
      firstAccount1DebitMovement.description,
      String.format(DEBIT_TRANSFER_DESCRIPTION, IntegrationTests.accountId2.get)
    )
    Assertions.assertEquals(
      firstAccount2CreditMovement.description,
      String.format(
        CREDIT_TRANSFER_DESCRIPTION,
        IntegrationTests.accountId1.get
      )
    )
  }

  @Order(8)
  @Test def assertTransfersWorkWithSuccessWithConcurrentParallelProcessing()
      : Unit = {
    IntStream
      .rangeClosed(0, 10)
      .parallel
      .forEach((i: Int) => {
        try assertTransfersWorkWithSuccess()
        catch {
          case ignored: Throwable =>
        }
      })

    val account1Movements = webClient.get
      .uri(
        "/api/v1/movements/" + IntegrationTests.accountId1.get + "?start=2000-10-31&end=3000-10-31&page=0&pageSize=100"
      )
      .exchange
      .expectStatus
      .is2xxSuccessful
      .expectBody(classOf[Array[Movement]])
      .returnResult
      .getResponseBody

    val account1Balance = webClient.get
      .uri("/api/v1/accounts/" + IntegrationTests.accountId1.get + "/balance")
      .exchange
      .expectStatus
      .is2xxSuccessful
      .expectBody(classOf[BalanceDTO])
      .returnResult
      .getResponseBody

    // based on the order of tests and the number of the intStream range transfer processes. The balance of account 1 must be zero.
    val account1BalanceOfMovements = retrieveTotalBalanceFromMovements(
      account1Movements
    )
    Assertions.assertEquals(account1Balance.balance, account1BalanceOfMovements)
    Assertions.assertEquals(
      IntegrationTests.expectedZero,
      account1BalanceOfMovements
    )

    val account2Movements = webClient.get
      .uri(
        "/api/v1/movements/" + IntegrationTests.accountId2.get + "?start=2000-10-31&end=3000-10-31&page=0&pageSize=100"
      )
      .exchange
      .expectStatus
      .is2xxSuccessful
      .expectBody(classOf[Array[Movement]])
      .returnResult
      .getResponseBody

    val account2Balance = webClient.get
      .uri("/api/v1/accounts/" + IntegrationTests.accountId2.get + "/balance")
      .exchange
      .expectStatus
      .is2xxSuccessful
      .expectBody(classOf[BalanceDTO])
      .returnResult
      .getResponseBody

    val account2BalanceOfMovements = retrieveTotalBalanceFromMovements(
      account2Movements
    )
    Assertions.assertEquals(account2Balance.balance, account2BalanceOfMovements)
    // based on the order of tests and the number of the intStream range transfer processes. The balance of account 2 must be 240.
    val expectedTotal = BigDecimal.valueOf(240).setScale(2, RoundingMode.FLOOR)
    Assertions.assertEquals(expectedTotal, account2BalanceOfMovements)
  }

  private def retrieveTotalBalanceFromMovements(movements: Array[Movement]) =
    util.Arrays
      .stream(movements)
      .map((movement: Movement) => {
        if (movement.`type` == MovementType.CREDIT) movement.amount
        else movement.amount.negate
      })
      .reduce(BigDecimal.ZERO, _ add _)
      .setScale(2, RoundingMode.DOWN)

  private def createMovementDTO = new CreateMovementDTO(
    IntegrationTests.expectedAmount
  )

  private def createAccountDTO =
    new CreateAccountDTO(UUID.randomUUID.toString, BigDecimal.ZERO)
}

File: launcher/src/test/java/launcher/integration/IntegrationTests.java
package launcher.integration;


import application.dto.BalanceDTO;
import application.dto.CreateAccountDTO;
import application.dto.CreateMovementDTO;
import application.dto.TransferMovementDTO;
import domain.model.Account;
import domain.model.Movement;
import domain.model.MovementType;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.Arrays;
import java.util.UUID;
import java.util.concurrent.atomic.AtomicReference;
import java.util.stream.IntStream;
import launcher.FinancialApplication;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.MethodOrderer;
import org.junit.jupiter.api.Order;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestMethodOrder;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.reactive.AutoConfigureWebTestClient;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.web.reactive.server.WebTestClient;
import org.springframework.web.reactive.function.BodyInserters;

import static domain.model.MovementDescription.CREDIT_TRANSFER_DESCRIPTION;
import static domain.model.MovementDescription.DEBIT_TRANSFER_DESCRIPTION;

@SpringBootTest(properties = {
        "spring.r2dbc.url=r2dbc:h2:mem:///./testdb",
        "spring.liquibase.url=jdbc:h2:mem:./testdb;DB_CLOSE_DELAY=-1"
})
@AutoConfigureWebTestClient
@ContextConfiguration(classes = {FinancialApplication.class})
@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
public class IntegrationTests {


    @Autowired
    private WebTestClient webClient;

    private static final AtomicReference<String> accountId1 = new AtomicReference<>();

    private static final AtomicReference<String> accountId2 = new AtomicReference<>();

    private static final BigDecimal expectedAmount = BigDecimal.TEN.setScale(2, RoundingMode.DOWN);

    private static final BigDecimal expectedZero = BigDecimal.ZERO.setScale(2, RoundingMode.DOWN);

    @Order(1)
    @Test
    public void createAccountsWithSuccess() {
        final var accountId1id = webClient.post()
                .uri("/api/v1/accounts")
                .contentType(MediaType.APPLICATION_JSON)
                .body(BodyInserters.fromValue(createAccountDTO()))
                .exchange()
                .expectStatus()
                .isCreated()
                .expectBody(Account.class)
                .returnResult()
                .getResponseBody()
                .id();

        accountId1.set(accountId1id);

        final var accountId2Id = webClient.post()
                .uri("/api/v1/accounts")
                .contentType(MediaType.APPLICATION_JSON)
                .body(BodyInserters.fromValue(createAccountDTO()))
                .exchange()
                .expectStatus()
                .isCreated()
                .expectBody(Account.class)
                .returnResult()
                .getResponseBody()
                .id();

        accountId2.set(accountId2Id);
    }

    @Order(2)
    @Test
    public void createDepositsWithSuccess() {
        webClient.post()
                .uri("/api/v1/movements/" + accountId1.get() + "/deposit")
                .contentType(MediaType.APPLICATION_JSON)
                .body(BodyInserters.fromValue(createMovementDTO()))
                .exchange()
                .expectStatus()
                .is2xxSuccessful()
                .expectBody(Movement.class);

        webClient.post()
                .uri("/api/v1/movements/" + accountId2.get() + "/deposit")
                .contentType(MediaType.APPLICATION_JSON)
                .body(BodyInserters.fromValue(createMovementDTO()))
                .exchange()
                .expectStatus()
                .is2xxSuccessful()
                .expectBody(Movement.class);
    }

    @Order(3)
    @Test
    public void createWithdrawsWithSuccess() {
        webClient.post()
                .uri("/api/v1/movements/" + accountId1.get() + "/withdraw")
                .contentType(MediaType.APPLICATION_JSON)
                .body(BodyInserters.fromValue(createMovementDTO()))
                .exchange()
                .expectStatus()
                .is2xxSuccessful()
                .expectBody(Movement.class);

        webClient.post()
                .uri("/api/v1/movements/" + accountId2.get() + "/withdraw")
                .contentType(MediaType.APPLICATION_JSON)
                .body(BodyInserters.fromValue(createMovementDTO()))
                .exchange()
                .expectStatus()
                .is2xxSuccessful()
                .expectBody(Movement.class);
    }

    @Order(4)
    @Test
    public void retrieveCreditMovements() {
        final var account1Movements = webClient.get()
                .uri("/api/v1/movements/" + accountId1.get() + "/credit")
                .exchange()
                .expectStatus()
                .is2xxSuccessful()
                .expectBody(Movement[].class)
                .returnResult()
                .getResponseBody();

        final var account2Movements = webClient.get()
                .uri("/api/v1/movements/" + accountId2.get() + "/credit")
                .exchange()
                .expectStatus()
                .is2xxSuccessful()
                .expectBody(Movement[].class)
                .returnResult()
                .getResponseBody();

        Assertions.assertEquals(1, account1Movements.length);
        Assertions.assertEquals(1, account2Movements.length);

        Assertions.assertEquals(expectedAmount, account1Movements[0].amount());
        Assertions.assertEquals(expectedAmount, account2Movements[0].amount());
    }

    @Order(5)
    @Test
    public void retrieveDebitMovements() {
        final var account1Movements = webClient.get()
                .uri("/api/v1/movements/" + accountId1.get() + "/debit")
                .exchange()
                .expectStatus()
                .is2xxSuccessful()
                .expectBody(Movement[].class)
                .returnResult()
                .getResponseBody();

        final var account2Movements = webClient.get()
                .uri("/api/v1/movements/" + accountId2.get() + "/debit")
                .exchange()
                .expectStatus()
                .is2xxSuccessful()
                .expectBody(Movement[].class)
                .returnResult()
                .getResponseBody();

        Assertions.assertEquals(1, account1Movements.length);
        Assertions.assertEquals(1, account2Movements.length);

        Assertions.assertEquals(expectedAmount, account1Movements[0].amount());
        Assertions.assertEquals(expectedAmount, account2Movements[0].amount());
    }

    @Order(6)
    @Test
    public void assertAccountsBalances() {
        final var account1Balance = webClient.get()
                .uri("/api/v1/accounts/" + accountId1.get() + "/balance")
                .exchange()
                .expectStatus()
                .is2xxSuccessful()
                .expectBody(BalanceDTO.class)
                .returnResult()
                .getResponseBody();

        final var account2Balance = webClient.get()
                .uri("/api/v1/accounts/" + accountId2.get() + "/balance")
                .exchange()
                .expectStatus()
                .is2xxSuccessful()
                .expectBody(BalanceDTO.class)
                .returnResult()
                .getResponseBody();

        Assertions.assertEquals(expectedZero, account1Balance.balance());
        Assertions.assertEquals(expectedZero, account2Balance.balance());
    }

    @Order(7)
    @Test
    public void assertTransfersWorkWithSuccess() {
        // deposit 10 bucks in each account
        createDepositsWithSuccess();

        final var transferMovement = new TransferMovementDTO(
                accountId2.get(),
                expectedAmount
        );

        webClient.post()
                .uri("/api/v1/movements/" + accountId1.get() + "/transfer")
                .contentType(MediaType.APPLICATION_JSON)
                .body(BodyInserters.fromValue(transferMovement))
                .exchange()
                .expectStatus()
                .is2xxSuccessful()
                .expectBody(Movement.class);

        final var account1Balance = webClient.get()
                .uri("/api/v1/accounts/" + accountId1.get() + "/balance")
                .exchange()
                .expectStatus()
                .is2xxSuccessful()
                .expectBody(BalanceDTO.class)
                .returnResult()
                .getResponseBody();

        final var account2Balance = webClient.get()
                .uri("/api/v1/accounts/" + accountId2.get() + "/balance")
                .exchange()
                .expectStatus()
                .is2xxSuccessful()
                .expectBody(BalanceDTO.class)
                .returnResult()
                .getResponseBody();

        Assertions.assertEquals(expectedAmount.add(expectedAmount), account2Balance.balance());
        Assertions.assertEquals(expectedZero, account1Balance.balance());

        final var account2CreditMovements = webClient.get()
                .uri("/api/v1/movements/" + accountId2.get() + "/credit")
                .exchange()
                .expectStatus()
                .is2xxSuccessful()
                .expectBody(Movement[].class)
                .returnResult()
                .getResponseBody();

        final var account1DebitMovements = webClient.get()
                .uri("/api/v1/movements/" + accountId1.get() + "/debit")
                .exchange()
                .expectStatus()
                .is2xxSuccessful()
                .expectBody(Movement[].class)
                .returnResult()
                .getResponseBody();

        final var firstAccount2CreditMovement = account2CreditMovements[0];
        final var firstAccount1DebitMovement = account1DebitMovements[0];
        Assertions.assertEquals(firstAccount1DebitMovement.amount(), firstAccount2CreditMovement.amount());

        Assertions.assertEquals(firstAccount1DebitMovement.description(), String.format(DEBIT_TRANSFER_DESCRIPTION, accountId2.get()));
        Assertions.assertEquals(firstAccount2CreditMovement.description(), String.format(CREDIT_TRANSFER_DESCRIPTION, accountId1.get()));
    }

    @Order(8)
    @Test
    public void assertTransfersWorkWithSuccessWithConcurrentParallelProcessing() {

        IntStream.rangeClosed(0, 10)
                .parallel()
                .forEach(i -> {
                    try {
                        assertTransfersWorkWithSuccess();
                    } catch (final Throwable ignored) {
                    }
                });

        final var account1Movements = webClient.get()
                .uri("/api/v1/movements/" + accountId1.get() + "?start=2000-10-31&end=3000-10-31&page=0&pageSize=100")
                .exchange()
                .expectStatus()
                .is2xxSuccessful()
                .expectBody(Movement[].class)
                .returnResult()
                .getResponseBody();

        final var account1Balance = webClient.get()
                .uri("/api/v1/accounts/" + accountId1.get() + "/balance")
                .exchange()
                .expectStatus()
                .is2xxSuccessful()
                .expectBody(BalanceDTO.class)
                .returnResult()
                .getResponseBody();


        // based on the order of tests and the number of the intStream range transfer processes. The balance of account 1 must be zero.
        final var account1BalanceOfMovements = retrieveTotalBalanceFromMovements(account1Movements);
        Assertions.assertEquals(account1Balance.balance(), account1BalanceOfMovements);
        Assertions.assertEquals(expectedZero, account1BalanceOfMovements);

        final var account2Movements = webClient.get()
                .uri("/api/v1/movements/" + accountId2.get() + "?start=2000-10-31&end=3000-10-31&page=0&pageSize=100")
                .exchange()
                .expectStatus()
                .is2xxSuccessful()
                .expectBody(Movement[].class)
                .returnResult()
                .getResponseBody();

        final var account2Balance = webClient.get()
                .uri("/api/v1/accounts/" + accountId2.get() + "/balance")
                .exchange()
                .expectStatus()
                .is2xxSuccessful()
                .expectBody(BalanceDTO.class)
                .returnResult()
                .getResponseBody();

        final var account2BalanceOfMovements = retrieveTotalBalanceFromMovements(account2Movements);
        Assertions.assertEquals(account2Balance.balance(), account2BalanceOfMovements);

        // based on the order of tests and the number of the intStream range transfer processes. The balance of account 2 must be 240.
        final var expectedTotal = BigDecimal.valueOf(240).setScale(2, RoundingMode.FLOOR);
        Assertions.assertEquals(expectedTotal, account2BalanceOfMovements);

    }

    private BigDecimal retrieveTotalBalanceFromMovements(final Movement[] movements) {
        return Arrays.stream(movements)
                .map(movement -> {
                    if (movement.type().equals(MovementType.CREDIT)) {
                        return movement.amount();
                    } else {
                        return movement.amount().negate();
                    }
                }).reduce(BigDecimal.ZERO, BigDecimal::add)
                .setScale(2, RoundingMode.DOWN);
    }

    private CreateMovementDTO createMovementDTO() {
        return new CreateMovementDTO(
                expectedAmount
        );
    }

    private CreateAccountDTO createAccountDTO() {
        return new CreateAccountDTO(
                UUID.randomUUID().toString(),
                BigDecimal.ZERO
        );
    }


}
